### PR TITLE
Initial Refactor of ursula prometheus metrics work from #1923

### LINKED ISSUE #2058: Iterate code from #1923
Related to #1923

### PR DESCRIPTION
**Pull request by derekpierre**
Related to #1923 

Closes #2058.

Follow-up #2091.

### CODE REVIEW

File: nucypher/characters/lawful.py [hunk 1]
```diff
@@ -1106,11 +1103,8 @@ def run(self,
             worker: bool = True,
             pruning: bool = True,
             interactive: bool = False,
-            prometheus: bool = False,
             start_reactor: bool = True,
-            metrics_port: int = None,
-            metrics_prefix: str = 'ursula',
-            metrics_listen_address: str = ''
+            prometheus_config: PrometheusMetricsConfig = None,
```
Comment (KPrasch) [inline nucypher/characters/lawful.py:1107]: Hmm - this needs a second look.  all of the other inputs to this function are feature switches as booleans.  Is this a special case?

File: nucypher/utilities/prometheus.py [hunk 2]
```diff
@@ -1,344 +0,0 @@
-"""
```
Comment (KPrasch) [inline nucypher/utilities/prometheus.py:1]: Im impressed you managed to dissolve these modules!

File: nucypher/utilities/prometheus/metrics.py [hunk 3]
```diff
@@ -0,0 +1,313 @@
...
+    from twisted.web.server import Site
+
+    metrics_prefix = prometheus_config.metrics_prefix
+    metrics_collectors = create_metrics_collectors(ursula, metrics_prefix)
+    # initalized collectors
+    node_metrics = {}
+    for collector in metrics_collectors:
+        node_metrics.update(collector.initialize(metrics_prefix=metrics_prefix, registry=ursula.registry))
+
+    # TODO: was never used
```
Comment (KPrasch) [inline nucypher/utilities/prometheus/metrics.py:133]: Let's follow up with the authors intention here.

File: nucypher/utilities/prometheus/metrics.py [hunk 4]
```diff
@@ -0,0 +1,313 @@
...
+    # "requests_counter": Counter(f'{metrics_prefix}_http_failures', 'HTTP Failures', ['method', 'endpoint']),
+
+    # Scheduling
+    metrics_task = task.LoopingCall(collect_prometheus_metrics,
+                                    metrics_collectors=metrics_collectors,
+                                    node_metrics=node_metrics)
+    metrics_task.start(interval=10, now=False)  # TODO: make configurable
```
Comment (KPrasch) [inline nucypher/utilities/prometheus/metrics.py:140]: Heh, might need a running list of configurations. It does not need to anything near comprehensive, but perhaps some high-level tuning knobs.

File: nucypher/utilities/prometheus/metrics.py [hunk 5]
```diff
@@ -0,0 +1,313 @@
...
+    collectors.append(EventMetricsCollector(
+        event_name='Deposited',
+        event_args_config={
+            "value": (Gauge, f'{metrics_prefix}_worklock_deposited_value', 'Deposited value')
+        },
+        argument_filters={"sender": ursula.checksum_address},
+        contract_agent=worklock_agent))
+
+    # Claimed
+    collectors.append(EventMetricsCollector(
```
Comment (KPrasch) [inline nucypher/utilities/prometheus/metrics.py:272]: this is looking pretty tidy but I am still wondering if it is possible to move this data generation to a (static?)method on each collector class.

File: tests/integration/cli/actions/test_auth_actions.py [hunk 6]
```diff
@@ -153,7 +153,7 @@ def test_unlock_nucypher_keyring(mocker,
 
     # Setup
     # Do not test "real" unlocking here, just the plumbing
-    unlock_spy = mocker.patch.object(NucypherKeyring, 'unlock', reeturn_value=True)
```
Comment (KPrasch) [inline tests/integration/cli/actions/test_auth_actions.py:156]: whoops!

File: tests/unit/test_prometheus.py [hunk 7]
```diff
@@ -1,23 +1,48 @@
...
+ along with nucypher.  If not, see <https://www.gnu.org/licenses/>.
+"""
+
 from __future__ import unicode_literals
 
+import json
 import sys
 import time
-import json
+import unittest
```
Comment (KPrasch) [inline tests/unit/test_prometheus.py:23]: Unused?

File: tests/unit/test_prometheus.py [hunk 8]
```diff
@@ -1,23 +1,48 @@
...
 
 from prometheus_client import (
     CollectorRegistry, Counter, Enum,
```
Comment (KPrasch) [inline tests/unit/test_prometheus.py:26]: Newlines! :-)

File: tests/unit/test_prometheus.py [hunk 9]
```diff
@@ -1,23 +1,48 @@
...
 
-from nucypher.utilities.json_metrics_export import JSONMetricsResource
+from nucypher.utilities.prometheus.metrics import JSONMetricsResource
+from nucypher.utilities.prometheus.metrics import PrometheusMetricsConfig
+
+TEST_PREFIX = 'test_prefix'
 
-from prometheus_client.core import GaugeHistogramMetricFamily, Timestamp
 
-if sys.version_info < (2, 7):
```
Comment (KPrasch) [inline tests/unit/test_prometheus.py:16]: :confused:

File: nucypher/characters/lawful.py (hunk 1 above)
Comment (KPrasch) [reply nucypher/characters/lawful.py:1107]: True.  The configuration needs to happen at an earlier stage to be consistent.

**Comment by derekpierre** [comment]
@KPrasch , @vepkenez - as of commit https://github.com/nucypher/nucypher/pull/2069/commits/6a578f034a67c3a6eac4fece13842b70f93d95ad we should be on par with the original functionality of the p2p code. Take a look - feel free to use it for nodes!

**Comment by derekpierre** [comment]
Once we agree on the architecture, we can start adding tests.

**Comment by derekpierre** [comment]
@KPrasch , @vepkenez let me know if you have any architectural thoughts regarding this PR. If not we can move on to adding unit tests.

No use writing unit tests if the architecture is going to significantly change. To get the code updated sooner rather than later, I could leave this PR as is, and add testing in a separate PR - wdyt?

**Comment by KPrasch** [comment]
> @KPrasch , @vepkenez let me know if you have any architectural thoughts regarding this PR. If not we can move on to adding unit tests.

We've been over this together a few times including a walkthrough - You've got my support of the design improvements here!  

I am also in support of merging this, and continueing to iterate in a follow up PR.  We'll get some live testing feedback earlier this way.

**Comment by KPrasch** [comment]
Rewor

**Pull request merged**